File: python/problems/leetcode/0992_subarrays_with_k_different_integers/solution.py
"""LeetCode 992: Subarrays with K Different Integers."""

from __future__ import annotations


class Solution:
    def subarraysWithKDistinct(self, nums: list[int], k: int) -> int:
        def at_most(k_limit: int) -> int:
            if k_limit < 0:
                return 0
            count: dict[int, int] = {}
            left = 0
            res = 0
            for right, x in enumerate(nums):
                count[x] = count.get(x, 0) + 1
                while len(count) > k_limit:
                    y = nums[left]
                    count[y] -= 1
                    if count[y] == 0:
                        del count[y]
                    left += 1
                res += right - left + 1
            return res

        return at_most(k) - at_most(k - 1)


if __name__ == "__main__":
    assert Solution().subarraysWithKDistinct([1, 2, 1, 2, 3], 2) == 7
    assert Solution().subarraysWithKDistinct([1, 2, 1, 3, 4], 3) == 3
    assert Solution().subarraysWithKDistinct([1, 1, 1], 1) == 6
    print("0992 OK")

File: cpp/problems/leetcode/0992_subarrays_with_k_different_integers/solution.cpp
// LeetCode 992: Subarrays with K Different Integers
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
    int atMost(const vector<int>& nums, int k) {
        if (k < 0) return 0;
        unordered_map<int, int> cnt;
        int left = 0, res = 0;
        for (int right = 0; right < (int)nums.size(); ++right) {
            ++cnt[nums[right]];
            while ((int)cnt.size() > k) {
                int y = nums[left++];
                if (--cnt[y] == 0) cnt.erase(y);
            }
            res += right - left + 1;
        }
        return res;
    }

public:
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        return atMost(nums, k) - atMost(nums, k - 1);
    }
};

int main() {
    vector<int> a{1, 2, 1, 2, 3};
    assert(Solution().subarraysWithKDistinct(a, 2) == 7);
    vector<int> b{1, 2, 1, 3, 4};
    assert(Solution().subarraysWithKDistinct(b, 3) == 3);
    vector<int> c{1, 1, 1};
    assert(Solution().subarraysWithKDistinct(c, 1) == 6);
    cout << "0992 OK" << endl;
    return 0;
}
